lib/auth/auth_dtls.c

```diff
@@ -0,0 +1,959 @@
+#define DTLS_HEADER_BYTES           (sizeof(struct dtls_packet_hdr))
+
+#define AUTH_DTLS_MIN_TIMEOUT       (10000u)
+#define ATUH_DTLS_MAX_TIMEOUT       (30000u)
+#define AUTH_DTLS_HELLO_WAIT_MSEC   (15000u)
+
+
+/**
+ * Header identifying a DTLS packet (aka datagram).  Unlike TLS, DTLS packets
+ * must be forwarded to Mbedtls as one or more complete packets.  TLS is
+ * design to handle an incoming byte stream.
+ */
+#pragma pack(push, 1)
+struct dtls_packet_hdr {
+	uint16_t sync_bytes;    /* use magic number to identify header */
+	uint16_t packet_len;    /* size of DTLS datagram */
+};
+#pragma pack(pop)
+
+
+
+/**
+ * Mbed context, one context per DTLS connection.
+ */
+struct mbed_tls_context {
+	bool in_use;
+
+	mbedtls_entropy_context entropy;
+	mbedtls_ctr_drbg_context ctr_drbg;
+	mbedtls_ssl_context ssl;
+	mbedtls_ssl_config conf;
+	mbedtls_x509_crt cacert;
+	mbedtls_x509_crt device_cert;
+	mbedtls_pk_context device_private_key;
+	mbedtls_timing_delay_context timer;
+	mbedtls_ssl_cookie_ctx cookie_ctx;
+
+	/* Temp buffer used to assemble full frame when sending. */
+	uint8_t temp_dtlsbuf[CONFIG_MBEDTLS_SSL_MAX_CONTENT_LEN];
+
+	/* cookie used for DTLS */
+	uint8_t cookie[AUTH_DTLS_COOKIE_LEN];
+};
+
+/**
+ * List of Mbed instances
+ */
+static struct mbed_tls_context tlscontext[MAX_MBEDTLS_CONTEXT];
+
+
+/* ===================== local functions =========================== */
+
+/**
+ * Get a free Mbed context to use with an authentication instance.
+ *
+ * @return Pointer to context, NULL if no free context available.
+ */
+static struct mbed_tls_context *auth_get_mbedcontext(void)
+{
+	// use mutex lock to protect accessing list
+	for (int cnt = 0; cnt < MAX_MBEDTLS_CONTEXT; cnt++) {
+
+		if (!tlscontext[cnt].in_use) {
+			tlscontext[cnt].in_use = true;
+			return &tlscontext[cnt];
+		}
+	}
+
+	return NULL;
+}
+
+/**
+ * Free a Mbed context.
+ *
+ * @param mbed_ctx Pointer to context.
+ */
+static void auth_free_mbedcontext(struct mbed_tls_context *mbed_ctx)
+{
+	mbed_ctx->in_use = false;
+
+	/* Free any MBed tls resources */
+	mbedtls_x509_crt_free(&mbed_ctx->cacert);
+	mbedtls_x509_crt_free(&mbed_ctx->device_cert);
+	mbedtls_pk_free(&mbed_ctx->device_private_key);
+	mbedtls_ssl_free(&mbed_ctx->ssl);
+	mbedtls_ssl_config_free(&mbed_ctx->conf);
+	mbedtls_ctr_drbg_free(&mbed_ctx->ctr_drbg);
+	mbedtls_entropy_free(&mbed_ctx->entropy);
+}
+
+/**
+ * Initialize Mbed context.
+ *
+ * @param mbed_ctx Pointer to context.
+ */
+static void auth_init_context(struct mbed_tls_context *mbed_ctx)
+{
+	mbedtls_ssl_init(&mbed_ctx->ssl);
+	mbedtls_ssl_config_init(&mbed_ctx->conf);
+	mbedtls_x509_crt_init(&mbed_ctx->cacert);
+	mbedtls_x509_crt_init(&mbed_ctx->device_cert);
+	mbedtls_pk_init(&mbed_ctx->device_private_key);
+	mbedtls_ssl_cookie_init(&mbed_ctx->cookie_ctx);
+	mbedtls_ctr_drbg_init(&mbed_ctx->ctr_drbg);
+	mbedtls_entropy_init(&mbed_ctx->entropy);
+
+	sys_rand_get(mbed_ctx->cookie, sizeof(mbed_ctx->cookie));
```

Comment:
don't you need cryptographically secure random numbers ? See `sys_csrand_get`

Reply:
Added new function `auth_get_random()` in auth_xport_common.c to handle SOCs with and without hardware based random number generators.

Reply:
In general, this code probably shouldn't be supported on SOCs that don't have hardware based random numbers. I wouldn't go out of the way to make it work, since the end result won't be secure.